ASGI middleware/plugin. (#84)
honeybadger-io/honeybadger-python: README.md

    return (a/b) #This will be reported
```

> ### [ASGI](https://asgi.readthedocs.io/en/latest/).

Reviewer: ```suggestion
### ASGI
```

> A generic ASGI middleware plugin is available for initializing and configuring Honeybadger: [`honeybadger.contrib.asgi`](./honeybadger/contrib/asgi.py).

Reviewer: ```suggestion
A generic [ASGI](https://asgi.readthedocs.io/en/latest/) middleware plugin is available for initializing and configuring Honeybadger: [`honeybadger.contrib.asgi`](./honeybadger/contrib/asgi.py).
```

The general pattern for these cases is wrapping your ASGI application with a middleware:

```python
from honeybadger import contrib

asgi_application = someASGIApplication()
asgi_application = contrib.ASGIHoneybadger(asgi_application)
```

> You can pass configuration parameters (or *additional* configuration parameters) as keyworkd arguments at plugin's initialization:

Reviewer: ```suggestion
You can pass configuration parameters (or *additional* configuration parameters) as keyword arguments at plugin's initialization:
```

```python
from honeybadger import contrib

asgi_application = someASGIApplication()
asgi_application = contrib.ASGIHoneybadger(asgi_application, api_key="<your-api-key>", params_filters=["sensible_data"])

```

Or you may want to initialize Honeybadger before your application, and then just registering the plugin/middleware:

```python
from honeybadger import honeybadger, contrib

honeybadger.configure(api_key='<your-api-key>')
some_possibly_failing_function()  # you can track errors happening before your plugin initialization.
asgi_application = someASGIApplication()
asgi_application = contrib.ASGIHoneybadger(asgi_application)
```

> ### [FastAPI](https://fastapi.tiangolo.com/).

Reviewer: ```suggestion
### FastAPI
```

> FastAPI is based on Starlette, a ASGI application.

Reviewer: ```suggestion
[FastAPI](https://fastapi.tiangolo.com/) is based on Starlette, an ASGI application.
```

You use Honeybadger's ASGI middleware on these types of applications.

```python

from fastapi import FastAPI
from honeybadger import contrib

app = FastAPI()
app.add_middleware(contrib.ASGIHoneybadger)
```

You can pass additional keyword paramters, too:

```python
from fastapi import FastAPI
from honeybadger import honeybadger, contrib

honeybadger.configure(api_key="<your-api-key>")
app = FastAPI()
app.add_middleware(contrib.ASGIHoneybadger, params_filters=["dont-include-this"])
```

#### FastAPI advanced usage.

> Consuming the request body in a ASGI application's middleware is [problematic and discouraged](https://github.com/encode/starlette/issues/495#issuecomment-494008175). This is the reason why request body data won't be sent to the web UI.

Reviewer: ```suggestion
Consuming the request body in an ASGI application's middleware is [problematic and discouraged](https://github.com/encode/starlette/issues/495#issuecomment-494008175). This is the reason why request body data won't be sent to the web UI.
```

FastAPI allows overriding the logic used by the `Request` and `APIRoute` classes, by [using custom `APIRoute` classes](https://fastapi.tiangolo.com/advanced/custom-request-and-route/). This gives more control over the request body, and makes it possible to send request body data along with honeybadger notifications. 

A custom API Route is available at [`honeybadger.contrib.fastapi`](./honeybadger/contrib/fastapi):

```python
from fastapi import FastAPI, APIRouter
from honeybadger import honeybadger, contrib

honeybadger.configure(api_key="<your-api-key>")
app = FastAPI()
app.router.route_class = contrib.HoneybadgerRoute

router = APIRouter(route_class=contrib.HoneybadgerRoute)

```

> ### [Starlette](https://www.starlette.io/).

Reviewer: ```suggestion
### Starlette
```

> You can configure Honeybadger to work with Starlette just like in any other ASGI framework.

Reviewer: ```suggestion
You can configure Honeybadger to work with [Starlette](https://www.starlette.io/) just like in any other ASGI framework.
```
